lib/worker.js

```diff
@@ -0,0 +1,135 @@
+ * @param {function} opts.done Callback to execute when the job has successfully
+ *   been completed.
+ * @param {boolean} [opts.runNow] Whether or not to run the job immediately,
+ *   defaults to `true`.
+ */
+function Worker (opts) {
+  // managed required fields
+  var fields = [
+    'queue',
+    'task',
+    'job',
+    'done'
+  ];
+  fields.forEach(function (f) {
+    if (!exists(opts[f])) {
+      throw new Error(f + ' is required for a Worker');
+    }
+  });
+
+  // manage field defaults
+  fields.push('runNow');
+  opts = pick(opts, fields);
+  defaults(opts, {
+    runNow: true,
+    // not necessarily user opts
+    attempt: 0,
+    retryDelay: process.env.WORKER_MIN_RETRY_DELAY
+  });
+
+  // put all opts on this
+  assign(this, opts);
+
+  log.info({ queue: this.queue, job: this.job }, 'Worker created');
+
+  if (this.runNow) {
+    this.run();
+  }
+}
+
+/**
+ * Factory method for creating new workers. This method exists to make it easier
+ * to unit test other modules that need to instantiate new workers.
+ * @see Worker
+ * @param {object} opts opts
+ * @returns {Worker} New Worker
+ */
+Worker.create = function (opts) {
+  return new Worker(opts);
+};
+
+/**
+ * Runs the worker. If the task for the job fails, then this method will retry
+ * the task (with an exponential backoff) a number of times defined by the
+ * environment of the process.
+ * @returns {promise} Promise resolved once the task succeeds or fails.
+ */
+Worker.prototype.run = function () {
+  var baseLogData = {
+    queue: this.queue,
+    job: this.job
+  };
+  return Promise.resolve().bind(this)
+    .then(function checkAttemptCounter () {
+      this.attempt++;
+      // Check if we've exceeded the maximum number of retries
+      if (this.attempt > process.env.WORKER_MAX_RETRIES) {
+        // TODO(bryan): add negative acknowledgements here when they are ready
+        var errData = {
+          job: this.job,
+          attempts: this.attempt
+        };
+        throw new TaskMaxRetriesError(this.queue, errData);
+      }
+    })
+    .then(function runTheTask () {
+      var attemptData = { attempt: this.attempt };
+      log.info(assign({}, baseLogData, attemptData), 'Running task');
+      return this.task(this.job);
+    })
+    .then(function successDone (result) {
+      log.info(assign({}, baseLogData, { result: result }), 'Job complete');
+      return this.done();
+    })
+    // if it's a known type of error, we can't accomplish the task
+    .catch(TaskFatalError, TaskMaxRetriesError, function knownErrDone (err) {
+      error.log(err);
+      // If we encounter a fatal error we should no longer try to schedule
+      // the job.
+      return this.done();
+    })
+    // TODO(bryan): do we want to do something special for TaskError?
+    .catch(function unknownErrRetry () {
+      var attemptData = { nextAttemptDelay: this.retryDelay };
+      log.warn(assign({}, baseLogData, attemptData), 'Task failed, retrying');
```

Comment:
I don't know how I feel about this style of `assign` logging. I almost feel like you have to do more work to figure out what is going on (by scrolling back up to the top of the function, etc.) than it would take if we simply included all the log fields explicitly on the statement.

Furthermore, it makes it harder to start jamming way too much information into any individual log line (as it becomes immediately apparent at the point of the log entry).